File: 232_implement_queue_using_stacks/MyQueue.ts
class MyQueue {
  input: number[];
  output: number[];

  constructor() {
    this.input = [];
    this.output = [];
  }

  push(x: number): void {
    this.input.push(x);
  }

  pop(): number {
    this.peek();
    return this.output.pop();
  }

  peek(): number {
    if (this.output.length === 0) {
      while (this.input.length) {
        this.output.push(this.input.pop());
      }
    }

    return this.output[this.output.length - 1];
  }

  empty(): boolean {
    return this.output.length === 0 && this.input.length === 0;
  }
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * var obj = new MyQueue()
 * obj.push(x)
 * var param_2 = obj.pop()
 * var param_3 = obj.peek()
 * var param_4 = obj.empty()
 */

File: 232_implement_queue_using_stacks/MyQueue.go
type MyQueue struct {
    input []int
    output []int
}


func Constructor() MyQueue {
    return MyQueue{}
}


func (this *MyQueue) Push(x int)  {
    this.input = append(this.input, x)
}


func (this *MyQueue) Pop() int {
    val := this.Peek()
    this.output = this.output[:len(this.output) - 1]
    return val
}


func (this *MyQueue) Peek() int {
    if len(this.output) == 0 {
        for len(this.input) > 0 {
            this.output = append(this.output, this.input[len(this.input) - 1])
            this.input = this.input[:len(this.input) - 1]
        }
    }
    return this.output[len(this.output) - 1]
}


func (this *MyQueue) Empty() bool {
    return len(this.input) == 0 && len(this.output) == 0
}


/**
 * Your MyQueue object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(x);
 * param_2 := obj.Pop();
 * param_3 := obj.Peek();
 * param_4 := obj.Empty();
 */